### PR TITLE
fix @boegel's remarks ;-)

### PR DESCRIPTION
**Pull request by boegel**
for https://github.com/hpcugent/vsc-base/pull/185

### CODE REVIEW

File: test/generaloption.py [hunk 1]
```diff
@@ -855,11 +869,15 @@ def test_option_as_value(self):
         self._match_testoption1_sysexit(['--store', '-b'],
                                         "Value '-b' is also a valid option")
 
+        # same for --store=-b
+        self._match_testoption1_sysexit(['--store=-b'],
```
Comment (stdweird) [inline test/generaloption.py]: why would this fail? this has to work since it's unambigous

Comment (boegel) [inline test/generaloption.py]: yes, indeed, that's also why my modified stuff doesn't pass the tests, will push another commit to fix this

**Comment by boegel** [comment]
@stdweird: updated, please check again, I removed the bogus test and implement support to recognise `-sb` etc. as a likely typo

File: lib/vsc/utils/generaloption.py [hunk 2]
```diff
@@ -488,7 +489,7 @@ def __init__(self, *args, **kwargs):
         self.environment_arguments = None
         self.commandline_arguments = None
 
-    def is_value_a_commandline_option(self, value, index=None):
+    def is_value_a_commandline_option(self, opt, value, index=None):
         """
         Determine if value is/could be an option passed via the commandline.
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: add something describing opt

File: lib/vsc/utils/generaloption.py [hunk 3]
```diff
@@ -497,28 +498,38 @@ def is_value_a_commandline_option(self, value, index=None):
 
         The method tests for possible ambiguity on the commandline when the parser
         interprets the argument following an option as a value, whereas it is far more likely that
-        it is (intended as) an option.
+        it is (intended as) an option; --longopt=value is never considered ambiguous, regardless of the value.
         """
         # Values that are/could be options that are passed via
-        # --longopt=value or -sVALUE are not a problem.
+        # --longopt=value is not a problem.
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: add `only`

File: lib/vsc/utils/generaloption.py [hunk 4]
```diff
@@ -497,28 +498,38 @@ def is_value_a_commandline_option(self, value, index=None):
...
+        cmdline_index = None
         try:
             cmdline_index = self.commandline_arguments.index(value)
         except ValueError:
-            # There is no ambiguity if the value is not passed
-            # as standalone argument via commandline
-            return None
+            # no index found for value, so not a stand-alone value
+            if opt.startswith('--'):
+                # --longopt=value is unambigouos
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: add `only`

File: lib/vsc/utils/generaloption.py [hunk 5]
```diff
@@ -497,28 +498,38 @@ def is_value_a_commandline_option(self, value, index=None):
...
+                return None
 
         if index is None:
             # index of last parsed arg in orig_rargs via remainder of rargs
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: `orig_rargs` -> `commandline_arguments` (i missed that somehow)

File: lib/vsc/utils/generaloption.py [hunk 6]
```diff
@@ -432,9 +432,10 @@ class ExtOptionParser(OptionParser):
     VALUES_CLASS = Values
     DESCRIPTION_DOCSTRING = False
 
-    ALLOW_OPTION_AS_VALUE = False # exact match for option as value
-    ALLOW_DASH_AS_VALUE = False # any value starting with a '-'
-    ALLOW_TYPO_AS_VALUE = True # value with similarity score from difflib.get_close_matches
+    ALLOW_OPTION_NAME_AS_VALUE = False  # exact match for option name (without the '-') as value
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: the name of the boolean is not correct, as this is only relevant for short options

Comment (boegel) [inline lib/vsc/utils/generaloption.py]: no it's not, using `-base` as a value is also an issue if `--base` is an option, see tests

File: lib/vsc/utils/generaloption.py [hunk 7]
```diff
@@ -497,28 +498,38 @@ def is_value_a_commandline_option(self, value, index=None):
...
             index = len(self.commandline_arguments) - len(self.rargs) - 1
 
-        if index != cmdline_index:
+        if cmdline_index is not None and index != cmdline_index:
             # This is not the value you are looking for
             return None
 
+        if not self.ALLOW_OPTION_NAME_AS_VALUE:
+            if '-%s' % value in self._short_opt or '-%s' % value in self._long_opt:
+                return "'-%(val)s' is a valid option, so using '%(val)s' as value is likely a typo" % {'val': value}
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: remove the `so using '%(val)s' as value is likely a typo` it's not required (and none of the other error messages have it)

File: lib/vsc/utils/generaloption.py [hunk 8]
```diff
@@ -497,28 +498,38 @@ def is_value_a_commandline_option(self, value, index=None):
...
             # index of last parsed arg in orig_rargs via remainder of rargs
             index = len(self.commandline_arguments) - len(self.rargs) - 1
 
-        if index != cmdline_index:
+        if cmdline_index is not None and index != cmdline_index:
             # This is not the value you are looking for
             return None
 
+        if not self.ALLOW_OPTION_NAME_AS_VALUE:
+            if '-%s' % value in self._short_opt or '-%s' % value in self._long_opt:
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: just do `'-%s' % value in self._short_opt + self._long_opt:`

File: test/generaloption.py [hunk 9]
```diff
@@ -838,14 +847,27 @@ def test_option_as_value(self):
...
+        self._match_testoption1_sysexit(['--store', 'b'],
+                                        "'-b' is a valid option, so using 'b' as value is likely a typo")
+        self._match_testoption1_sysexit(['--store', '-base'],
+                                        "'--base' is a valid option, so using '-base' as value is likely a typo")
+        self._match_testoption1_sysexit(['-s', '-base'],
+                                        "'--base' is a valid option, so using '-base' as value is likely a typo")
+        self._match_testoption1_sysexit(['-s-base'],
+                                        "'--base' is a valid option, so using '-base' as value is likely a typo")
+
+        #self._match_testoption1_sysexit(['-s-b'],
```
Comment (stdweird) [inline test/generaloption.py]: remove if unneeded

File: test/generaloption.py [hunk 10]
```diff
@@ -838,14 +847,27 @@ def test_option_as_value(self):
...
-        # valid ways of specifying '-b' as a value to --store/-s
+        # only valid way of specifying '-b' as a value to --store
         topt = TestOption1(go_args=['--store=-b'])
         self.assertEqual(topt.options.store, '-b')
         self.assertFalse(topt.options.base)  # remain False (default value)
 
-        topt = TestOption1(go_args=['-s-b'])  # equivalent to --store=-b
-        self.assertEqual(topt.options.store, '-b')
-        self.assertFalse(topt.options.base)  # remain False (default value)
+        # when -b/--base is an option, the following are not accepted, since they're likely not what intended
```
Comment (stdweird) [inline test/generaloption.py]: add tests with `ALLOW_OPTION_NAME_AS_VALUE` explicitly set to True or False, and test the default value

File: lib/vsc/utils/generaloption.py [hunk 11]
```diff
@@ -488,37 +489,48 @@ def __init__(self, *args, **kwargs):
...
+                return "'-%(val)s' is a valid option, so using '%(val)s' as value is likely a typo" % {'val': value}
+
         # First test the exact problem where the value is an option
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: not really `First` anymore

File: lib/vsc/utils/generaloption.py [hunk 12]
```diff
@@ -488,37 +489,48 @@ def __init__(self, *args, **kwargs):
...
         if (not self.ALLOW_OPTION_AS_VALUE) and (value in self._long_opt or value in self._short_opt):
```
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: also here `value in self._long_opt + self._short_opt`

File: lib/vsc/utils/generaloption.py (hunk 6 above)
Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: ok

**Comment by stdweird** [comment]
@boegel thx a lot

**Pull request merged**